ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Connection.scala

```diff
@@ -140,14 +141,21 @@ private[h2] class H2Connection[F[_]](
     } >>
       H2Connection.KillWithoutMessage().raiseError
 
+  private[this] def mapRefForEach(f: H2Stream[F] => F[Unit]): F[Unit] = {
+    @tailrec
+    def go(iter: Iterator[H2Stream[F]], acc: F[Unit]): F[Unit] =
+      if (iter.hasNext) go(iter, acc *> f(iter.next())) else acc
```

Comment:
Isn't iter.hasNext an effect?

Reply:
An iterator is a mutable structure, and each `next()` pops the next element. Since `*>` is strict, this would force the evaluation of `f(iter.next())`, which would itself force the evaluation of `iter.next()`, all before doing the recursive call to `go`. Mutability is contained within the method, and different calls to the method should create different iterators.